service/nodes.go

```diff
@@ -103,82 +128,120 @@ type connectorNode struct {
 		case component.DataTypeMetrics:
-			n.Component, err = builder.CreateMetricsToTraces(ctx, set, fanoutConsumer)
+			conn, err = builder.CreateMetricsToTraces(ctx, set, fanoutConsumer)
 		case component.DataTypeLogs:
-			n.Component, err = builder.CreateLogsToTraces(ctx, set, fanoutConsumer)
+			conn, err = builder.CreateLogsToTraces(ctx, set, fanoutConsumer)
 		}
 	case component.DataTypeMetrics:
 		var consumers []consumer.Metrics
 		for _, next := range nexts {
-			metricsConsumer, ok := next.(consumer.Metrics)
-			if !ok {
-				return fmt.Errorf("next component is not a metrics consumer: %s", n.componentID)
-			}
-			consumers = append(consumers, metricsConsumer)
+			consumers = append(consumers, next.(consumer.Metrics))
 		}
 		fanoutConsumer := fanoutconsumer.NewMetrics(consumers)
-		switch n.exprPipelineType {
+		switch exprPipelineType {
 		case component.DataTypeTraces:
-			n.Component, err = builder.CreateTracesToMetrics(ctx, set, fanoutConsumer)
+			conn, err = builder.CreateTracesToMetrics(ctx, set, fanoutConsumer)
 		case component.DataTypeMetrics:
-			n.Component, err = builder.CreateMetricsToMetrics(ctx, set, fanoutConsumer)
+			conn, err = builder.CreateMetricsToMetrics(ctx, set, fanoutConsumer)
 		case component.DataTypeLogs:
-			n.Component, err = builder.CreateLogsToMetrics(ctx, set, fanoutConsumer)
+			conn, err = builder.CreateLogsToMetrics(ctx, set, fanoutConsumer)
 		}
 	case component.DataTypeLogs:
 		var consumers []consumer.Logs
 		for _, next := range nexts {
-			logsConsumer, ok := next.(consumer.Logs)
-			if !ok {
-				return fmt.Errorf("next component is not a logs consumer: %s", n.componentID)
-			}
-			consumers = append(consumers, logsConsumer)
+			consumers = append(consumers, next.(consumer.Logs))
 		}
 		fanoutConsumer := fanoutconsumer.NewLogs(consumers)
-		switch n.exprPipelineType {
+		switch exprPipelineType {
 		case component.DataTypeTraces:
-			n.Component, err = builder.CreateTracesToLogs(ctx, set, fanoutConsumer)
+			conn, err = builder.CreateTracesToLogs(ctx, set, fanoutConsumer)
 		case component.DataTypeMetrics:
-			n.Component, err = builder.CreateMetricsToLogs(ctx, set, fanoutConsumer)
+			conn, err = builder.CreateMetricsToLogs(ctx, set, fanoutConsumer)
 		case component.DataTypeLogs:
-			n.Component, err = builder.CreateLogsToLogs(ctx, set, fanoutConsumer)
+			conn, err = builder.CreateLogsToLogs(ctx, set, fanoutConsumer)
 		}
 	}
-	return err
+	return
+}
+
+var _ consumerNode = &capabilitiesNode{}
+
+// Every pipeline has a "virtual" capabilities node immediately after the receiver(s).
+// There are two purposes for this node:
+// 1. Present aggregated capabilities to receivers, such as whether the pipeline mutates data.
+// 2. Present a consistent "first consumer" for each pipeline.
+// The nodeID is derived from "pipeline ID".
+type capabilitiesNode struct {
+	nodeID
+	pipelineID component.ID
+	baseConsumer
+	consumer.Capabilities
```

Comment:
We build the graph in such a way that each `receiverNode` has edges _only_ to one or more `capabilitiesNode`.

Before we build a receiver component, we call `componentGraph.nextConsumers` to get the `[]capabilitiesNode` for that receiver. (Returned as `[]baseConsumer` but actually is always `[]capabilitiesNode` for a receiver.)

Then in `buildReceiver`, we build a new `fanoutconsumer` that takes into account `Capabilities.MutatesData`. This `fanoutconsumer` is given to the new receiver component as its next consumer.

All of the above is true for connectors too - they always emit to `fanoutconsumer([]capabilitiesNode)` which accounts for `MutatesData`.

Separately, `HandleZPages` describes each pipeline including whether it `MutatesData`. This can easily be determined from the `capabilitiesNode` in each pipeline. (Equivalent to `builtPipeline.lastConsumer`.)

Reply:
@bogdandrutu and I discussed briefly offline and determined that there was indeed a problem with the capabilities node. I've believe I've addressed it in [this commit](https://github.com/open-telemetry/opentelemetry-collector/pull/7045/commits/63976de092281c9aceee32ee2ec88717732ced99).